### PR TITLE
Implement DOM overlays for expressing conventions + build files in the terms of a DOM

### PR DESCRIPTION
**Pull request by h0tk3y**

### CODE REVIEW

**Comment by h0tk3y** [comment]
@bot-gradle test this

**Comment by h0tk3y** [comment]
@bot-gradle test this

**Comment by h0tk3y** [comment]
@bot-gradle test this

File: platforms/core-configuration/declarative-dsl-core/src/main/kotlin/org/gradle/internal/declarativedsl/dom/operations/overlay/OverlayNodeOrigin.kt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+package org.gradle.internal.declarativedsl.dom.operations.overlay
+
+import org.gradle.internal.declarativedsl.dom.DeclarativeDocument
+
+
+/**
+ * In a document overlay, tells where a node comes from.
+ */
+sealed interface OverlayNodeOrigin {
```
Comment (eskatos) [inline platforms/core-configuration/declarative-dsl-core/src/main/kotlin/org/gradle/internal/declarativedsl/dom/operations/overlay/OverlayNodeOrigin.kt:25]: 👍

File: platforms/core-configuration/declarative-dsl-core/src/test/kotlin/org/gradle/internal/declarativedsl/dom/operations/overlay/DocumentOverlayTest.kt [hunk 2]
```diff
@@ -0,0 +1,446 @@
...
+import org.gradle.internal.declarativedsl.dom.resolution.DocumentWithResolution
+import org.gradle.internal.declarativedsl.dom.resolution.documentWithResolution
+import org.gradle.internal.declarativedsl.parsing.ParseTestUtil
+import org.gradle.internal.declarativedsl.schemaBuilder.schemaFromTypes
+import kotlin.test.Test
+import kotlin.test.assertEquals
+
+
+internal
+object DocumentOverlayTest {
```
Comment (eskatos) [inline platforms/core-configuration/declarative-dsl-core/src/test/kotlin/org/gradle/internal/declarativedsl/dom/operations/overlay/DocumentOverlayTest.kt:38]: 👍

**Review by eskatos — APPROVED** [review]
LGTM!

**Pull request merged**